text/0000-maybe-dangling.md

```diff
@@ -0,0 +1,268 @@
+- [Yoke](https://github.com/unicode-org/icu4x/issues/3696) and [Yoke again](https://github.com/unicode-org/icu4x/issues/2095) (the first needs opting-out of `dereferenceable` for the yoke, the latter needs opting-out of `noalias` for both yoke and cart)
+- [ouroboros](https://github.com/joshua-maros/ouroboros/issues/88)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+To handle situations like this, Rust has a special type called `MaybeDangling<P>`:
+references and boxes in `P` do *not* have to be dereferenceable or follow aliasing guarantees.
+This applies inside nested references/boxes inside `P` as well.
+They still have to be non-null and aligned, and it has to at least be *possible* that there exists valid data behind that reference (i.e., `MaybeDangling<&!>` is still invalid).
+Also note that safe code can still generally assume that every `MaybeDangling<P>` it encounters is a valid `P`, but within unsafe code this makes it possible to store data of arbitrary type without making reference guarantees (this is similar to `ManuallyDrop`).
+In other words, `MaybeDangling<P>` is entirely like `P`, except that the rules that relate to the contents of memory that pointers in `P` point to (dereferencability and aliasing restrictions) are suspended when the pointers are not being actively used.
+You can think of the `P` as being "suspended" or "inert".
+
+The `ManuallyDrop<T>` type internally wraps `T` in a `MaybeDangling`.
+
+This means that the first example is actually fine:
+the dangling `Box` was passed inside a `ManuallyDrop`, so there is no UB.
+
+The 2nd example can be fixed by passing the closure in a `MaybeDangling`:
+```rust
+// Argument is passed as `MaybeDangling` since we might actually keep 
+// it around after its lifetime ends (at which point the caller can 
+// start dropping memory it points to).
+fn thread(control: ..., closure: MaybeDangling<impl FnOnce() + 'lifetime>) {
+    closure.into_inner()();
+    control.signal_done();
+    // A lot of time can pass here.
+}
+```
+
+The 3rd example can be fixed by storing the `buffer` inside a `MaybeDangling`, which disables its aliasing requirements:
+
+```rust
+struct OwningRef<U, T: StableDeref<Target=U>> {
+    buffer: MaybeDangling<T>,
+    ref_: NonNull<U>, // conceptually borrows from `buffer`.
+}
+```
+
+As long as the `buffer` field is not used, the pointer stored in `ref_` will remain valid.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The standard library contains a type `MaybeDangling<P>` that is safely convertible with `P` (i.e., the safety invariant is the same), and that has all the same niches as `P`, but that does allow passing around dangling boxes and references within unsafe code.
```

Comment:
How is it possible to have both (1) `MaybeDangling<P>` and `P` are safely convertible and (2) `P` holds more guarantees than `MaybeDangling<P>`? If they are convertible, one could safely pass say `MaybeDangling<&'a mut T>` to a `fn(&'a mut T, ...)` which itself is compiled with noalias assumption, and result in undefined behaviors. If `P -> MaybeDangling<P>` is safe and `MaybeDangling<P> -> P` is unsafe, then what's the difference between `MaybeDangling` and `MaybeUninit`?

I also disagree with the disagreement by @RalfJung 's https://github.com/rust-lang/rfcs/pull/3336#issuecomment-1368935572:
>> MaybeDangling as currently specified by the RFC does not change safety invariants, as it implements safe >> Deref conversion like ManuallyDrop does. I argue that this is likely undesirable, and MaybeDangling should instead make it unsafe to reactivate the wrapped type and its invariants.
>
> I think that would make MaybeDangling mostly useless -- one can already use MaybeUninit for something like that. I consider not having to use unsafe code to get the inner data back out of the no-dangling wrapper to be a feature. For example, this would have avoided adding unsafe code in https://github.com/rust-lang/rust/pull/102589.

No. It's intrinsically unsafe to have `fn(SomeType<'some_life>)` where `'some_life` *may* ends in the middle of the function invocation depending on its logic. In that case (example 2 of this RFC), the validity is controlled by the runtime logic and is impossible to check at compile time (eg. cannot prevent you from "accidentally" swapping `closure();` and `control.signal_done();`), thus guarantees should be made by programmer, `unsafe` is required, with or without `MaybeDangling`. Then I don't really see the advantage of introducing another `MaybeDangling` than `MaybeUninit`.

The dereferenceability of `ManuallyDrop` in example 1, I think, is a separate issue. It's a choice between UB on state observed and UB on state exists.

Reply:
> How is it possible to have both (1) MaybeDangling<P> and P are safely convertible and (2) P holds more guarantees than MaybeDangling<P>? 

There are [validity invariants and safety invariants](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html). The safety invariant of `P` and `MaybeDangling<P>` are identical, but the validity invariants are not.

> Then I don't really see the advantage of introducing another MaybeDangling than MaybeUninit.

`MaybeUninit` is a terrible solution for example 2 since one has to call `drop` manually. That's fragile and error-prone. I think it is important that we have a way to not lose automatic drop but still pass around "opaque arbitrary data" without making extra claims about its memory validity.

Once we have implicit `drop`, there's no reason to make `into_inner` unsafe, since `drop` will already implicitly convert to the inner type without unsafe code.